### DISCUSSION TITLE
S-Parameter simulations using Meep

### DISCUSSION THREAD

**Discussion by jan-david-fischbach** [gdsfactory/gdsfactory#310] [Q&A] [answered]
When running S-Parameter simulations with meep (e.g. by using gmeep's `gm.write_sparameters_meep`) I frequently end up with |S_11| larger than 1 (or larger than 0dB if expressed that way). What are possible causes for this behavior? Using the S-parameter extraction based on Lumerical this behavior was not observed.
Best Regards
JD

**Comment by simbilod** [comment]
Hi JD,

I have seen such discrepancies in the past where the default port sizes would draw ports that overlapped with many core regions, and therefore did not monitor the proper modes. If this is what's going on, there are function arguments to offset the ports to mitigate this.

If you attach your script with your geometry I can take a look.

Best,

Simon

**Reply by jan-david-fischbach** [reply]
[S-Params_gt1.zip](https://github.com/gdsfactory/gdsfactory/files/8597039/S-Params_gt1.zip)
I hope you can recreate it this way.

This is my S11 for example:
![s11](https://user-images.githubusercontent.com/92856893/166108285-5bf1e7b2-91c7-4e06-b1f8-9b98909c0820.png)

from this setup:
![setup](https://user-images.githubusercontent.com/92856893/166108479-5541319c-8288-477d-8d55-17f606c3c111.png)

**Reply by simbilod** [reply]
Hi JD,

I cannot reproduce your issue -- all my S-parameters for your script are below 0dB. However, my gdsfactory was complaining about "ge" being present, so I toggled that off by mapping it to Si. It seemed like your stack didn't contain germanium anyway, but correct me if I am wrong.

![image](https://user-images.githubusercontent.com/46427609/166179248-1fc3024e-58e7-462f-b23e-858ee8bcecb2.png)
![image](https://user-images.githubusercontent.com/46427609/166179260-6a5ec26e-f8ed-4e04-ab5e-273b02c51b1f.png)


I did notice your monitors are not quite into the input and output waveguides. So I also repeated you simulation while adding xmargins, extending the port waveguides, and offsetting the port monitors like so:

```
df = gm.write_sparameters_meep(c, 
                                xmargin_left=1, 
                                xmargin_right=1, 
                                ymargin_top=3, 
                                ymargin_bot=3, 
                                extend_ports_length=4, 
                                port_source_offset=0.5,
                                port_monitor_offset=0.5,
                                resolution=100, 
                                layer_stack=stack, 
                                filepath='/home/simbil/debugging/df_res100_padded.csv', 
                                dirpath="/home/simbil/debugging/", 
                                overwrite=True, 
                                animate=False,,)
```
This gives the simulation below


https://user-images.githubusercontent.com/46427609/166179334-21b8ab2d-9461-4cd3-9e04-7452a7b7579c.mp4

![image](https://user-images.githubusercontent.com/46427609/166179291-f847d3b4-aa66-410f-a364-51d209701c09.png)
![image](https://user-images.githubusercontent.com/46427609/166179314-46b84c9f-4764-4edb-8550-6a6d4f44e560.png)


I saw no difference here, but since the eigenmode coefficients are computed from the modal cross section of the port, it is good to ensure they are well embedded into the input and output waveguides.

**Reply by simbilod** [reply]
This was all done with resolution=100. I am a bit puzzled by my very low reflection compared to what is qualitatively seen on the animation.

**Reply by jan-david-fischbach** [reply]
Hi Simon,
you are safe to ignore the Germanium (map it to **SiO2**), as it is not used in the simulations yet. As you have mapped it to Si this would represent a very different geometry. But I think it should not influence the simulation due to the 2D simulation region not intersecting with the Ge-Layer. In your second plot S21m is larger than 1 for small wavelength, which is equivalent to being larger than 0dB, is that wrong? 

> I did notice your monitors are not quite into the input and output waveguides

I did not notice that, thank you for pointing it out!! That might very much be the issue! I somehow missed the gap under tha port...

> I saw no difference here, but since the eigenmode coefficients are computed from the modal cross section of the port, it is good to ensure they are well embedded into the input and output waveguides.

From what I see there is a difference in your simulation results: S21m is now < 1 for all lambda
Btw: left to right is S21, right not S12, correct?

Thanks a lot for the help. I'll se whether that helps to sort things out.
Best Regards
JD

**Reply by jan-david-fischbach** [reply]
> This was all done with resolution=100. I am a bit puzzled by my very low reflection compared to what is qualitatively seen on the animation.

The S-Params I get out do represent this relatively high reflection. I am also puzzeled by the fact, that the S-Params we get out vary so drastically.

btw: Do 2D simulations with resolution 100 take ages for you guys too? Or do I have something setup incorrectly? Mine is currently running for more than 10 minutes and has not finished calculating the forward direction...

**Reply by simbilod** [reply]
After mapping Ge to SiO2 I do get something more realistic:

![image](https://user-images.githubusercontent.com/46427609/166243311-5d5ff61c-49c8-46d7-9cd2-648e2b3282b2.png)
![image](https://user-images.githubusercontent.com/46427609/166243341-3c7c0e78-9f02-43f1-9712-6e5928cb1a0d.png)

**Reply by jan-david-fischbach** [reply]
Interesting. I thought that the Germanium layer was disregarded, because it was not intersected by the simulation region.

**Comment by joamatab** [comment]
agreed with Simon

I think Lumerical has a normalization to make sure the total power of the Sparameters adds to 100%

**Comment by simbilod** [comment]
@Jan-David-Black yes serial simulations are excruciatingly slow, which is why we also made an interface to Parallel MEEP. It's a slightly different command, and note the "cores" argument:

```
df_file = gm.write_sparameters_meep_mpi(c, 
                                xmargin_left=1, 
                                xmargin_right=1, 
                                ymargin_top=3, 
                                ymargin_bot=3, 
                                extend_ports_length=4, 
                                port_source_offset=0.5,
                                port_monitor_offset=0.5,
                                resolution=100, 
                                layer_stack=stack, 
                                filepath='/home/simbil/debugging/df_res100_padded_GetoSiO2.csv', 
                                dirpath="/home/simbil/debugging/", 
                                overwrite=True, 
                                animate=False,
                                cores=6,)
```
It also currently returns a filepath instead of the dataframe, so you do need to tweak the rest of the code too:

```
# Read the filepath into a dataframe before continuing
import pandas as pd
df = pd.read_csv(df_file)
gf.simulation.plot.plot_sparameters(df, keys=("s11m",))
```

We may change this at some point to make it more consistent.

**Reply by simbilod** [reply]
This takes ~250 seconds to complete with 2.1GHz Intel Xeon Cascadelake processors

**Reply by jan-david-fischbach** [reply]
I am going to try it out. But right now it is failing in a bit of a weird way: It prints 
```
mpirun -np 6 python <conda-env>/lib/python3.10/site-packages/gdslib/sp/temp/write_sparameters_meep_mpi.py
<conda-env>/lib/python3.10/site-packages/gdslib/sp/<device_name>_142408ee.csv
```
the file name that is returned does not exist :/

**Reply by simbilod** [reply]
Running MPI from Python is a bit of a hack, it needs to write a file and then execute it. It seems like you installed gdsfactory via conda, so it tries to create the temporary file there, and possibly can't find it afterwards.

Try giving it the temporary script location manually, like below:

```
df_file = gm.write_sparameters_meep_mpi(c, 
                                xmargin_left=1, 
                                xmargin_right=1, 
                                ymargin_top=3, 
                                ymargin_bot=3, 
                                extend_ports_length=4, 
                                port_source_offset=0.5,
                                port_monitor_offset=0.5,
                                resolution=100, 
                                layer_stack=stack, 
                                filepath='/home/simbil/debugging/df_res100_padded_GetoSiO2.csv', # Fileeath for output, choose your own
                                dirpath="/home/simbil/debugging/",  # Dirpath for output, choose your own
                                overwrite=True, 
                                animate=False,
                                cores=6,
                                temp_dir='/home/simbil/debugging/', # Dirpath for MPI script, choose your own
                                temp_file_str='/home/simbil/debugging/', # Filename for MPI script choose your own
)
```

**Reply by jan-david-fischbach** [reply]
That unfortunately did not help. I tried executing mpirun directly but ran into this error:
```
Fatal error in internal_Init_thread: Other MPI error, error stack:
internal_Init_thread(60)...........: MPI_Init_thread(argc=0x0, argv=0x0, required=3, provided=0x30a86d7ec) failed
MPII_Init_thread(209)..............: 
MPID_Init(75)......................: 
init_world(190)....................: channel initialization failed
MPIDI_CH3_Init(84).................: 
MPID_nem_init(313).................: 
MPID_nem_tcp_init(175).............: 
MPID_nem_tcp_get_business_card(397): 
GetSockInterfaceAddr(370)..........: gethostbyname failed, jfischbach-MacBook-Pro (errno 0)
Fatal error in internal_Init_thread: Other MPI error, error stack:
internal_Init_thread(60)...........: MPI_Init_thread(argc=0x0, argv=0x0, required=3, provided=0x302f537ec) failed
MPII_Init_thread(209)..............: 
MPID_Init(75)......................: 
init_world(190)....................: channel initialization failed
MPIDI_CH3_Init(84).................: 
MPID_nem_init(313).................: 
MPID_nem_tcp_init(175).............: 
MPID_nem_tcp_get_business_card(397): 
GetSockInterfaceAddr(370)..........: gethostbyname failed, jfischbach-MacBook-Pro (errno 0)
```
(times six)
I seem to have some problems with my mpi install

**Reply by simbilod** [reply]
Did you install MEEP with the mpi flag?

`conda install -c conda-forge pymeep=*=mpi_mpich_* -y`

or 

`mamba install -c conda-forge pymeep=*=mpi_mpich_* -y`

**Reply by jan-david-fischbach** [reply]
yep. except for the fact that I used
```
conda install -c conda-forge "pymeep=*=mpi_mpich_*" -y
```
(mind the quotes)
Due to problems with zsh

**Reply by simbilod** [reply]
Hum seems like a higher level issue

**Reply by jan-david-fischbach** [reply]
But I had another version of meep installed previously. I'll try to reinstall

**Reply by jan-david-fischbach** [reply]
I think I got it fixed: It was some problem with macos, as the hostname of the machine was not saved in /etc/hosts. Correcting that resulted in the simulation running, and finishing. Somehow the simulation result look different from the results obtained before using MPI. Now I cannot rerun these simulations however, as the kernel crashes when running the simulation (without MPI)

edit: It does not crash if I run it by 
```
mpirun python extract_s.py
```

**Reply by jan-david-fischbach** [reply]
Well, seemingly the crashing is just an issue with interactive python. These are the results obtained from the two commands:
```python
df = gm.write_sparameters_meep_mpi(c, 
                                ymargin=3, 
                                xmargin=1, 
                                resolution=20, 
                                layer_stack=stack, 
                                extend_ports_length=4, 
                                port_source_offset=0.5,
                                port_monitor_offset=0.5,
                                material_name_to_meep = {'si': 3.45, 'sio2': 1.45, 'ge': 1.45},
                                cores=6, overwrite=True
                                )

df_serial = gm.write_sparameters_meep(c, 
                                ymargin=3, 
                                xmargin=1, 
                                resolution=20, 
                                layer_stack=stack, 
                                extend_ports_length=4, 
                                port_source_offset=0.5,
                                port_monitor_offset=0.5,
                                material_name_to_meep = {'si': 3.45, 'sio2': 1.45, 'ge': 1.45},
                                overwrite=True
                                )
```
![Figure_1](https://user-images.githubusercontent.com/92856893/166263618-112d4c5b-7ecf-4148-b1bf-40735f81d9cf.png)
![Figure_1(serial)](https://user-images.githubusercontent.com/92856893/166263622-e280e731-b3de-4ab2-8f1f-bda990d2bdcb.png)

(in the same order)

**Reply by simbilod** [reply]
Hum very strange considering that the `_mpi` version uses the other one. Seems like a legitimate bug. I will look into it later. That'll teach me not to have enough tests...

**Reply by jan-david-fischbach** [reply]
Thanks for looking into it :)

**Reply by simbilod** [reply]
I've fixed the MPI issue, and I will do a PR shortly. In your case some of the reflection still seems to be over 1

![image](https://user-images.githubusercontent.com/46427609/167234847-1579821b-163e-4911-b276-306ab2d17e27.png)

**Reply by jan-david-fischbach** [reply]
Very nice, that you found and fixed the issue with MPI. I am still struggeling to understand the reflections > 1 though. For me it seems to reduce with grid resolution though. Thus I assume its some discretization artefact.

**Comment by smartalecH** [comment, marked as answer]
I would check out this:

https://meep.readthedocs.io/en/latest/FAQ/#why-are-my-reflectancetransmittance-values-less-than-zero-andor-greater-than-one

And this

https://meep.readthedocs.io/en/latest/FAQ/#how-do-i-compute-s-parameters

I think the meep offloading needs a bit of revamping. I plan to tackle this in the near future.
